### PR TITLE
implement mergesort

### LINKED ISSUE #25: Parameter -mergesort

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #25

### CODE REVIEW

**Comment by justinmueller05082000** [comment]
```
panic: runtime error: index out of range [2] with length 2

goroutine 1 [running]:
main.mergeSort(0xc000010240, 0x2, 0x2, 0x4f2800, 0xc000010240, 0x2, 0x2)
        /home/jmueller/go/src/github.com/justinmueller05082000/trainee-sort/main.go:191 +0x492
main.merge(0xc000010240, 0x2, 0x2, 0x4f2800, 0xc000010240, 0x2, 0x2, 0x1, 0xc000010240, 0x2)
        /home/jmueller/go/src/github.com/justinmueller05082000/trainee-sort/main.go:204 +0xe9
main.main()
        /home/jmueller/go/src/github.com/justinmueller05082000/trainee-sort/main.go:97 +0x97b
exit status 2

```

File: main.go [hunk 1]
```diff
@@ -167,3 +171,38 @@ func bubbleSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
...
+		}
+	}
+
+	fmt.Println(mergeSort(input[:mid], comp))
+	fmt.Println(mergeSort(input[mid:], comp))
+
+	return nil
+}
+
+func merge(input [][]byte, comp func(a, b []byte) int, data [][]byte) [][]byte {
```
Comment (Al2Klimov) [inline main.go]: What's this for?

File: main.go [hunk 2]
```diff
@@ -167,3 +171,38 @@ func bubbleSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
...
+
+	for i := 0; i < len(input); i++ {
+		if comp(input[i],input[i+1]) > 0 {
+			mergeDuplicate[i], mergeDuplicate[i+1] = input[i+1], input[i]
+			return input
+		}
+	}
+
+	fmt.Println(mergeSort(input[:mid], comp))
+	fmt.Println(mergeSort(input[mid:], comp))
```
Comment (Al2Klimov) [inline main.go]: You have to store mergeSort()s results into variables **and then** merge them into mergeDuplicate.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go [hunk 3]
```diff
@@ -94,15 +93,15 @@ func main() {
...
+		dataSplit = mergeSort(dataSplit, comparator)
 	} else {
 		bubbleSort(dataSplit, comparator)
 	}
 
 	dataJoin := bytes.Join(dataSplit, []byte{'\n'})
 
 	if *writingParameter != "" {
-		writeErr := ioutil.WriteFile(*writingParameter, append(dataJoin,'\n'), 0644)
+		writeErr := ioutil.WriteFile(*writingParameter, append(dataJoin, '\n'), 0644)
```
Comment (Al2Klimov) [inline main.go]: Not part of the task.

File: main.go [hunk 4]
```diff
@@ -173,36 +172,45 @@ func bubbleSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
 }
 
 func mergeSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
-	if len(input) < 2 {
+	var num = len(input)
+
+	if num == 1 {
```
Comment (Al2Klimov) [inline main.go]: Why not `len(input) < 2` anymore?

File: main.go [hunk 5]
```diff
@@ -173,36 +172,45 @@ func bubbleSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
...
-	for i := 0; i < len(input); i++ {
-		mergeDuplicate[i] = make([]byte, len(input))
-		for j := 0; j < len(input[0]); j++ {
-			mergeDuplicate[i][j] = input[i][j]
-		}
-	}
+	var (
+		left  = mergeSort(input[:middle], comp)
+		right = mergeSort(input[middle:], comp)
+	)
```
Comment (Al2Klimov) [inline main.go]: Can these two variables be avoided?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [x] git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)
* [x] git commit --amend --no-edit
* [x] git push -f origin feature/implement-mergesort-25

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol comparing -mergesort and no -mergesort.

**Comment by justinmueller05082000** [comment]
**Test protocol:** 

**toSort.txt**
```
5
1
3
65
53
3
341
23
11
```

**Case 1:**
**Terminal:**
`go run . toSort.txt`
**Output:**
```
1
11
23
3
3
341
5
53
65
```


**Case 2:**
**Terminal:**
`go run . -mergesort toSort.txt`
**Output:**
```
1
11
23
3
3
341
5
53
65
```

**Conclusion:**
Both results are equal.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**